mobile/mobile.go

```diff
@@ -2,57 +2,39 @@ package mobile
 
 import (
 	"context"
+	"fmt"
 
-	ipfslite "github.com/hsanjuan/ipfs-lite"
-	"github.com/ipfs/go-log"
-	crypto "github.com/libp2p/go-libp2p-crypto"
-	"github.com/multiformats/go-multiaddr"
 	"github.com/textileio/grpc-ipfs-lite/server"
+	"github.com/textileio/grpc-ipfs-lite/util"
 )
 
-// Start starts the mobile ipfs-lite peer and gRPC server
-func Start(datastorePath string) error {
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
-
-	log.SetLogLevel("*", "warn")
-
-	ds, err := ipfslite.BadgerDatastore(datastorePath)
-	if err != nil {
-		return err
-	}
-	priv, _, err := crypto.GenerateKeyPair(crypto.RSA, 2048)
-	if err != nil {
-		return err
-	}
-
-	listen, _ := multiaddr.NewMultiaddr("/ip4/0.0.0.0/tcp/4005")
+var (
+	cancel context.CancelFunc
+)
 
-	h, dht, err := ipfslite.SetupLibp2p(
-		ctx,
-		priv,
-		nil,
-		[]multiaddr.Multiaddr{listen},
-	)
+const ipfsPort = 4005
+const grpcPort = 10000
 
-	if err != nil {
-		return err
-	}
+// Start starts the mobile ipfs-lite peer and gRPC server
+func Start(datastorePath string, debug bool) (int, error) {
+	var ctx context.Context
+	ctx, cancel = context.WithCancel(context.Background())
 
-	lite, err := ipfslite.New(ctx, ds, h, dht, nil)
+	lite, err := util.NewPeer(ctx, datastorePath, ipfsPort, debug)
 	if err != nil {
-		return err
+		return 0, err
 	}
 
-	lite.Bootstrap(ipfslite.DefaultBootstrapPeers())
+	host := fmt.Sprintf("localhost:%d", grpcPort)
 
 	// TODO: run this in a goroutine, but need to get any error back out, but this blocks, so how do you know it is running with no error?
-	go server.StartServer(lite, "localhost:10000")
+	go server.StartServer(lite, host)
 
-	return nil
+	return grpcPort, nil
 }
 
 // Stop stops the embedded grpc and ipfs servers
 func Stop() {
 	server.StopServer()
+	cancel()
```

Comment:
@sanderpick this file illustrates something I don't yet understand about gomobile...

This file is bound by gomobile. It has a `var cancel` that is set in the `Start` method and then referenced at some later time here in the `Stop` method. Does gomobile essentially model this as a class with a instance var and maintain that state/reference internally?

Reply:
Like, I can't tell if this will work as is or if setting package level vars like this is meaningless in the context of an export gomobile package.

Reply:
This will probably work, but I think it would be cleaner to return an instance that has the start and stop method, like `go-textile/mobile`.

Reply:
I want to try it out like this to see how it feels on the native side. Will keep this in mind.